Adding functions to compute the visual width of east asian characters

DilsonHiga opened the pull request:
Implement function to compute the East Asian Width Property of codepoints as decribed in [issue 6](https://github.com/roc-lang/unicode/issues/6#issue-2268644676).

I am new not only to roc but to functional programming in general, so I expect this code to be unidiomatic. I am open to criticism and to fix any issues.

The results of the function have been tested against other open source libraries.

lukewilliamboswell commented on package/InternalEAW.roc:
```diff
@@ -0,0 +1,10 @@
+## WARNING This file is automatically generated. Do not edit it manually. ##
+module [eastAsianWidthProperty, eastAsianWidth]
+
+eastAsianWidth : U32 -> U32
```
Added a type annotation here so that the compiler will only generate one variant.

lukewilliamboswell commented on package/InternalEAWGen.roc:
```diff
@@ -0,0 +1,140 @@
+
+EawRange : (Str, Str, Str)
+
+main =
+    when Arg.list! |> List.get 1 is
+        Err _ -> Task.err (InvalidArguments "USAGE: roc run InternalEAWGen.roc -- path/to/package/")
+        Ok arg -> File.writeUtf8 "$(Helpers.removeTrailingSlash arg)/InternalEAW.roc" template
```
Upgraded to basic-cli 0.11.0, we can simplify this a lot. The other Gen scripts were written a while ago and a lot has changed since then. For example we can use the `Arg.list!` inline here which I think is nice.

lukewilliamboswell commented on package/InternalEAWGen.roc:
```diff
@@ -0,0 +1,140 @@
+import "data/EastAsianWidth-15.1.0.txt" as file : Str
+import Helpers
+
+EawRange : (Str, Str, Str)
```
Added an alias for this so it is easier to follow what the below functions are doing and add a type annotation. In general I think most of the top-level declarations should be annotated.

lukewilliamboswell commented on package/InternalEAWGen.roc:
```diff
@@ -0,0 +1,140 @@
+        |> Str.fromUtf8
+        |> Result.map \end -> (eawp, start, "0x$(end)")
+
+expect parseLine "0020           ; Na # Zs         SPACE" == Ok ("Na", "0x0020", "0x0020")
+expect parseLine "0025..0027     ; Na # Po     [3] PERCENT SIGN..APOSTROPHE" == Ok ("Na", "0x0025", "0x0027")
```
Added a couple of unit tests, mostly to help check my understanding of this code.

lukewilliamboswell commented on package/InternalEAWGen.roc:
```diff
@@ -0,0 +1,140 @@
+
+    List.append res.merged res.merging
+
+mergeOp : { merged : List EawRange, merging : EawRange }, EawRange -> { merged : List EawRange, merging : EawRange }
+mergeOp = \{ merged, merging }, currentValue ->
+    if merging.0 == currentValue.0 && (Helpers.hexStrToU32 merging.2) + 1 == (Helpers.hexStrToU32 currentValue.1) then
+        {
+            merged,
+            merging: (merging.0, merging.1, currentValue.2),
+        }
+    else
+        {
+            merged: List.append merged merging,
+            merging: currentValue,
+        }
+
+expect mergeOp { merged: [], merging: ("Na", "0x0020", "0x0020") } ("Na", "0x0021", "0x0021") == { merged: [], merging: ("Na", "0x0020", "0x0021") }
+expect mergeOp { merged: [], merging: ("Na", "0x0020", "0x0020") } ("Na", "0x0022", "0x0022") == { merged: [("Na", "0x0020", "0x0020")], merging: ("Na", "0x0022", "0x0022") }
```
Pulled this out into a separate helper and added some unit tests because I was finding it hard to follow.

lukewilliamboswell reviewed (changes requested):
This is looking good 😍 Thank you very much for tackling this.

Next I think we should add the EAW module to the package, and an example which uses the EAW, which a test script for CI.

Also, can you please sign your commits?

DilsonHiga commented:
Thank you for your review!

I added a commit in which I incorporated the functions for getting the EAW property and the width of code points into the CodePoint module, but I'm not sure if that is the way to do it.

I have also added some tests to both the generated module and it's generator.
